python_modules/dagster/dagster/_utils/source_position.py

```diff
@@ -0,0 +1,107 @@
+    line: int
+    col: int
+
+
+class SourcePosition(NamedTuple):
+    filename: str
+    start: LineCol
+    end: LineCol
+
+    def __str__(self):
+        return f"{self.filename}:{self.start.line}"
+
+
+class SourcePositionTreeNode(NamedTuple):
+    position: SourcePosition
+    children: Mapping[KeyPathSegment, "SourcePositionTreeNode"]
+
+    def __hash__(self) -> int:
+        return hash((self.position, tuple(sorted(self.children.items()))))
+
+    def lookup(self, key_path: KeyPath) -> Optional[SourcePosition]:
+        """Returns the source position of the descendant at the given path."""
+        if len(key_path) == 0:
+            return self.position
+        head, *tail = key_path
+        if head not in self.children:
+            return None
+        return self.children[head].lookup(tail)
+
+
+class ValueAndSourcePositionTreeNode(NamedTuple):
+    value: Any
+    tree_node: SourcePositionTreeNode
+
+
+class SourcePositionAndKeyPath(NamedTuple):
+    """Represents a source position and key path within a file.
+
+    Attributes:
+        key_path (KeyPath): The path of keys that lead to the current object, where each element in
+            the path is either a string (for dict keys) or an integer (for list indices).
+        source_position (Optional[SourcePosition]): The source position of the object in the
+            document, if available.
+    """
+
+    key_path: KeyPath
+    source_position: Optional[SourcePosition]
+
+
+class HasSourcePositionAndKeyPath:
+    _source_position_and_key_path: Optional[SourcePositionAndKeyPath] = None
+
+
+def populate_source_position_and_key_paths(
+    obj: Any,
+    tree_node: Optional[SourcePositionTreeNode],
+    key_path: KeyPath = [],
+) -> None:
+    """Populate the SourcePositionAndKeyPath for the given object and its children.
+
+    This function recursively traverses the object and its children, setting the
+    SourcePositionAndKeyPath on each object that subclasses HasSourcePositionAndKeyPath.
+
+    The SourcePositionAndKeyPath is set based on the provided tree node, which contains the source
+    position information for the object and its children.
+
+    Args:
+        obj (Any): The object to populate the source position and key path for.
+        tree_node (Optional[SourcePositionTreeNode]): The tree node containing the source
+            position information for the object and its children.
+        key_path (KeyPath): The path of keys that lead to the current object.
+    """
+    if isinstance(obj, HasSourcePositionAndKeyPath):
+        source_position = None
+        if tree_node is not None:
+            source_position = tree_node.position
+        source_position_and_key_path = SourcePositionAndKeyPath(key_path, source_position)
+        if (
+            obj._source_position_and_key_path is not None  # noqa: SLF001
+            and obj._source_position_and_key_path != source_position_and_key_path  # noqa: SLF001
+        ):
+            raise RuntimeError(
```

Comment:
What's the reason for this? Just to make sure we don't accidentally do duplicate work?

Reply:
Yeah just something has probably gone wrong if we end up here. Changed it to a `check.invariant`.